---
title: "Function and distribution analyses of virus-encoded hydrocarbon-degradation genes"
author:
  - name: Jinlong Ru
toc: false
number-sections: true
format: docx
bibliography: refs.bib
csl: viruses.csl
params:
  name: "98-manuscript"
---

Updated: `r format(Sys.time(), '%Y-%m-%d %H:%M:%S', tz = 'CET')` CET.


```{r setup}
#| message: false
#| include: false
library(here)
here::i_am(paste0(params$name, ".qmd"), uuid = "230ddaf3-12db-4d34-ba8f-0bdd42ff27cb")
projthis::proj_create_dir_target(params$name, clean = FALSE)
path_target <- projthis::proj_path_target(params$name)
path_source <- projthis::proj_path_source(params$name)
path_data <- path_source("00-raw", params$name)
dir.create(path_data, recursive = T)
```


## Abstract

Microbial bioremediation is a widely accepted approach for hydrocarbon degradation. It involves using bacteria to degrade hydrocarbon pollution in the environment. However, little is known about the role of viruses in the bioremediation process and hydrocarbon degradation. In this study, we analyzed viral proteins in public databases and found that viruses encode auxiliary metabolic genes (AMGs) that may be involved in various types of hydrocarbon degradation. We identified 57 viral genomes encoding high-quality hydrocarbon-degradation genes. These findings suggest that viral AMGs could contribute to the degradation of alkanes and aromatic hydrocarbons, which would be beneficial for biodegradation and viral propagation. We also constructed a comprehensive database to facilitate future screening and validation of these genes. This study provides the first insight into the existence of various hydrocarbon-degradation genes in viral genomes, their functions, and their distribution. It will open up new perspectives on the role of viruses in microbial ecology and global carbon cycling.

## Introduction

Hydrocarbon pollution is a significant source of environmental contamination, leading to adverse effects on both the environment and human health [@2022-Components_Pavlenko]. Bioremediation, which employs microorganisms to degrade hydrocarbons, has emerged as a promising and eco-friendly strategy for mitigating hydrocarbon pollution [@2017-Coliform_Khanafer;@2022-Car_Vidal-Verdu]. Despite the extensive research on the role of bacteria and fungi in hydrocarbon biodegradation [@2022-Biodegradation_Ghorbannezhad;@2014-Oil_Wang;@2019-Microbial_Patel;@2019-Metagenomic_Zhang;@2021-Combined_Zhang], the potential contribution of viruses to this process remains largely unexplored.

Recent studies have demonstrated that viruses often contain auxiliary metabolic genes (AMGs), which can contribute to the adaptive stress resistance of their bacterial hosts , augment host metabolism, and enhance viral fitness. For example, AMGs have been found to assist photosynthesis in cyanobacteria [@2004-Transfer_Lindell], aid in biofilm formation [@2015-Filamentous_Secor], and participate in microbial sulfur metabolism [@2021-Ecology_Kieft]. These findings raise the possibility that viruses might also contribute to hydrocarbon degradation via AMGs. However, in contrast to the known bacterial phyla capable of dissimilatory sulfur metabolism [@2021-Ecology_Kieft], fewer bacteria have been reported to degrade crude oil. To date, isolated or enriched anaerobic alkane degraders belong to two phyla: Proteobacteria and Firmicutes [@2016-High_Stagars], suggesting a limited range of oil degradation-related AMGs that could be acquired by bacteriophages. Furthermore, some rare AMGs are subject to turnover due to fluctuating natural selection pressures [@2016-Genomic_Crummett], indicating that despite the ubiquity of phage-carried AMGs, their specific context and function are under natural selection. These factors make the discovery of oil-degradation related AMGs in viral genomes particularly challenging.

In this study, we aimed to investigate the potential role of viruses in hydrocarbon biodegradation by identifying and characterizing virus-encoded hydrocarbon-degradation AMGs (vHYDEGs). We performed a systematic search for hydrocarbon-degradation genes in the Integrated Microbial Genomes/Viruses (IMG/VR) database and identified 595 putative vHYDEGs, of which 57 were classified as high quality. These vHYDEGs were grouped into 15 protein families involved in the initial key steps of hydrocarbon degradation, such as hydroxylation, carboxylation, and fumarate addition. Phylogenetic analyses revealed a diverse evolutionary history of these genes but also demonstrated their close relationships with known bacterial oil degraders. Protein structural analyses indicated that vHYDEGs showed high similarity to experimentally validated proteins, suggesting that they possess the same metabolic potential as their homologous genes in bacteria. All identified vHYDEGs and their annotations were integrated into the vHyDeg database to facilitate future research. Overall, our study provides new insights into the contribution of viruses to bioremediation and highlights the importance of considering the role of viruses in microbial ecology and bioremediation strategies.

## Materials and methods

### vHYDEG identification and classification

To identify viral proteins that participate the initial activation step of hydrocarbon degradation, we downloaded viral proteins from the Integrated Microbial Genomes/Viruses (IMG/VR v4) [@2022-IMG_Camargo] and PHROGs (v4) [@2021-PHROG_Terzian] database. Only clustered PHROGs proteins and proteins from hign-confidence genomes in IMG/VR were used. Protein sequences longer than 10,000 amino acids were removed. In total, more than 115 million proteins were obtained. Protein sequences were first annotated using hmmscan to search against CANT-HYD database, which is a curated database of hydrocarbon degradation genes (HYDEG). Each HYDEG family was represented by a profile hidden markov model (hydHMM), which has a "trusted" cutoff to determine if a query sequence can be confidently annotated for the function. Sequences above the trusted cutoff can be confidently assigned a function. Since the trusted cutoff of each hydHMM was determined by the full-length of that gene in non-viral genomes, using these cutoffs may underestimate the number of viral-encoded HYDEGs (vHYDEGs), because viruses often encode truncated versions of metabolic genes which can still perform function [@2020-Large_Chen]. Therefore, we defined a relaxed cutoff for each hydHMM using the following strategy. First, we downloaded all viral profile HMMs (vHMM) and their protein sequences from the PHROGs database, which contains viral protein families with remote homologous. Second, we annotated all protein sequences of each vHMM using hydHMMs. As protein families in the PHROGs database were high-quality viral proteins, we assume they were not homologous of any hydrocarbon-degradation genes and hydHMMs. Therefore, the highest bit-score (named as “bitscore_vmax”) of each targeted hydHMM was set as the relaxed cutoff for that hydHMM. To reduce false positive rates, the relaxed cutoff was set to 150 if the respective bitscore_vmax is smaller than 150. All relaxed cutoff were smaller than 150 and original “trusted” cutoff, indicating that there is no full-length proteins with hydrocarbon-degradation functions in the PHROGs database. Hits of IMG/VR proteins with a bit-score greater than the relaxed cutoff, e-value<0.00001, and viral genome length>5 kbp were considered as putative vHYDEGs. To increase accuracy of protein domain annotation, all putative vHYDEGs were further annotated using Pfam database [@2021-Pfam_Mistry].

Putative vHYDEGs were further classified into three groups. (1). A vHYDEG was classified as high-quality (HQ) if the trusted index > 0.5, and its respected genomes are not proviruses and has an estimated contamination ratio=0. The trusted_idx threshold is set as 0.5 is based on the fact that the confirmed bacteriophage-encoded methane degradation gene pBmoC [@2020-Large_Chen] has a trusted index equal to 0.56 by our annotation method. Noting that the threshold “0.5” is selected based only on pBmoC gene, this threshold is a rough setting and can be adjusted for specific protein of interest in the future. (2). A vHYDEG was classified as medium-quality (MQ) if the trusted index > 0.5, but respected genomes were classified as proviruses or has an estimated contamination ratio>0. (3). All other vHYDEGs were classified as low-quality (LQ). For each vHYDEG family, the one with highest trusted index score in HQ was selected as representative sequence (HQrep). In total, we got 362 LQ, 176 MQ and 57 HQ putative vHYDEGs. Six HQreps were used for down-stream structural, phylogenetic and genome analyses.

### vHYDEG protein structure prediction and comparison

Protein sturcture of six HQreps were predicted using ColabFold [@2022-ColabFold_Mirdita], which combines MMseqs2 [@2017-MMseqs2_Steinegger] for fast homology search and AlphaFold2 [@2021-Highly_Jumper] for accurate structure prediction. Predicted structures were then compared with pretein structures in UniProt [@2019-UniProt_TheUniProtConsortium] and RCSB PDB database [@2000-Protein_Berman] using foldseek [@2022-Foldseek_Kempen]. Structural alignment was evaluated using TM-score normalized by length of the reference protein [@2004-Scoring_Zhang]. Two structures was considered as a good match if the TM-score is above 0.5. Structural comparison results were manually checked to select the top hits that annotated as relative functions. The top hit in PDB database was choosen if its annotation is the same function as the query protein, otherwise, the top hit in UniProt database was choosen and the related Alphafold-predicted structures was selected from the AlphaFold Protein Structure Database (AlphaFold DB) [@2022-AlphaFold_Varadi] as target structure. Binding site and domain features were extracted from UniProt feature annotations, and visualized on the structural alignment of the query and target setructures using open source PyMOL ([https://github.com/schrodinger/pymol-open-source](https://github.com/schrodinger/pymol-open-source)).

### vHYDEG protein phylogenetic tree reconstruction

A phylogenetic analysis was used to further investigate the evolutionary origin of the representative vHYDEGs. Protein sequences of each HQrep was searched against the NCBI refseq_protein database [@2021-RefSeq_Li] using BLASTp with bitscore>50 and evalue<0.001 to recruit closely related sequences and to add non-viral context to the phylogenetic analyses. Non-viral hits in the top 100 blast hits were clustered by 90% sequence similarity using MMseqs2 [@2017-MMseqs2_Steinegger] to reduce redundancy. Viral hits were combined with the representative sequences of each cluster and corresponding HQ vHYDEGs, and used for build multiple sequence alignment (MSA) using MUSCLE [@2004-MUSCLE_Edgar]. Phylogenetic trees were then created with MSAs using IQ-TREE [@2015-IQTREE_Nguyen] with the “LG + G4” model. Visualization of the phylogenetic tree was created using ggtree [@2022-Ggtree_Xu] with the mVC metadata from IMG/VR database and taxonomy annotations.

### mVC genome annotation and taxonomy assignment

Genome sequences of the 57 HQ vHYDEGs were downloaded from IMG/VR database (v4). For consistency genomic annotations, we annotated them using Pharokka pipeline [@2022-Pharokka_Bouras], which uses PHANOTATE [@2019-PHANOTATE_McNair] to predict open reading frames, and PHROGs database [@2021-PHROG_Terzian] to annotate protein functions. The identififed protein sequences were further annotated using InterProScan [@2014-InterProScan_Jones] to get more broad function annotation. Since some viruses uses sulfur or nitrogen as electron donor during hydrocarbon degradation, we also annotated proteins using nitrogen metabolism genes [@2020-Potential_Gazitua] from KEGG database [@2021-KEGG_Kanehisa], and sulfur metabolism genes from a custom HMM database [@2018-Expanded_Anantharaman;@2022-Diversity_Wolf]. Genome annotations were visualized using pyCircos v0.3.0 ([https://github.com/ponnhide/pyCircos](https://github.com/ponnhide/pyCircos)).

Taxonomy of mVCs were assigned using two approaches that implemented in the ViroProfiler pipeline ([https://github.com/deng-lab/viroprofiler](https://github.com/deng-lab/viroprofiler)). One was the protein-sharing network analysis using vConTACT2 [@2019-Taxonomic_BinJang], which clustered mVCs with reference genomes based on their shared protein clusters. NCBI viral RefSeq (v211) was selected as the reference database. The second one was a protein-voting-based method using MMseqs2 taxonomy module [@2016-Fast_Menzel], which searched proteins of mVCs against proteins in NCBI viral RefSeq database, add then assign a lowest-common ancestor (LCA) taxonomy based on mojority voting of protein taxonomy in that mVC. The conflict taxonomic assignments of these two methods were manually checked to select the LCA taxonomy. Finally, the protein-sharing network was visualized using Cytoscape [@2003-Cytoscape_Shannon].

### Comparative Genomics Analysis

To compare the genomic context of vHYDEGs with related sequences, we downloaded genome of the top bacteria hit in the BLASTp results and compared it with corresponding annotated vHYDEG viral genomes using Clinker [@2021-Clinker_Gilchrist]. Clinker used MMseqs2 to cluster proteins in these genomes and then generated genome maps and linked genes belonging to the same protein family through links. Annotation of bacterial genomes were downloaded from NCBI in GenBank format. If there was no annotation, the FASTA file of genomes was downloaded and then annotated using the Bakta annotation pipeline [@2021-Bakta_Schwengers].

### vHyDeg database construction

To facilitate future research on vHYDEGs, we created a database called vHyDeg, which contains all vHYDEGs identified in this study. The vHyDeg database integrates information on vHYDEG names, comprehensive functional annotations, genomic and host information. Protein functional annotations from different databases are linked to the original database using hyperlinks. Genome annotations can be accessed using links to the IMG/VR database. Additionally, the database will be constantly updated with new vHYDEGs as they are discovered. The vHyDeg database can be accessed at [https://deng-lab.github.io/vhydeg](https://deng-lab.github.io/vhydeg).

## Results

### Viruses encode various hydrocarbon-degradation genes

We identified 595 putative vHYDEGs form IMG/VR database (v4), which contains 15 of 37 protein families in CANT-HYD database. Based on the quality classification threshold, 362 of them were classified as low quality (LQ), 176 medium quality (MQ) and 57 high quality (HQ). Detailed annotation of all vHYDEGs can be found in the vHyDeg database ([https://deng-lab.github.io/vhydeg](https://deng-lab.github.io/vhydeg)). Number of viral genomes of each MQ and HQ vHYDEG were shown in Fig1A. The 57 HQ vHYDEGs include 33 pBmoC, 13 CYP153, 6 AlkB, 2 AlmA, one LadA alpha subunit and one pBmoA (Table S1). AlkB, CYP153, LadA alpha subunit and AlmA are alkane hydroxylases which contribute to the first activation step of aerobic long-chain alkane degradation, while pBmoC and pBmoA are subunits of particulate methane monooxygenase (pMMO) that involved in oxidation of methane to methanol, the pBmoC gene has been reported to be identified from bacteriophage genomes [@2020-Large_Chen]. The rest nine vHYDEGs classified as MQ and LQ including different categories of hydrolases and alkane monooxygenases, and targeted on both middle length and long-chain alkane as well as the polycyclic aromatic hydrocarbons.

We found that the trusted index distribution of vHYDEGs was skewed towards low values (Fig1B), with only 17 (2.85%) has trusted index > 1, and 78 (13.1%) has trusted index > 0.5. Although viruses were believed to transfer genes between different hosts, the prevalence of low trusted index indicate that most vHYDEGs have low sequence similarity to known hydrocarbon-degradation enzymes, or contain domains for other metabolic functions. To identify these domains that were not included in the CANT-HYD proteins, we further annotated all vHYDEGs using Pfam database. The annotation results showed that many of them are involved in hydrocarbon degradation pathways(Fig1C). For example, top Pfam annotations include ring hydroxylating, molybdopterin oxidoreductase, and pyruvate formate lyase-like domains, which are all related to either aerobic or anaerobic hydrocarbon-degradation [@2022-Genomeresolved_Somee;@2016-Anaerobic_Rabus;@2019-Anaerobic_Liu]. Considering that viruses usually encode partial genes and recombined genes, we hypothesis that vHYDEGs may be involved in hydrocarbon degradation in different ways. However, further research is needed to validate this hypothesis.

### Viruses-encoded alkane hydroxylases are responsible for key steps of degradation

According to the function of HQ vHYDEGs, *AlkB*, *CYP153*, *ladA* and *AlmA* are coding alkane hydroxylases(AH). These evidences showing the potential involvement of bacteriophages in the initial step of aerobic alkane degradation. The AHs in bacteria who catalyse the first step of alkane biodegradation, from alkane to iso-ethanol, are the key enzymes of aerobic degradation of alkanes. AlkB and CYP450 family proteins are two common proteins of AH in bacteria, besides, the flavin-binding LC-alkane hydroxylase (*AlmA*) and thermophilic soluble LC-alkane hydroxylase (*LadA*) were also proved to be involved in the hydroxylation of alkanes[@2013-Enzymes_Shao;@2020-Integrated_Xu], and plenty of bacteria were found to contain multiple types or copies of AH[@2014-Diverse_Nie;@2015-Occurrence_Viggor]. Previous studies confirmed that in order to survive in harsh environments such as salty crude oil, or facilitate the interspecies cooperation during crude oil degradation, bacteria evolves a series of corresponding functionals protein or metabolic systems through horizontal gene transfer[@2022-Auxiliary_Heyerhoff;@2019-Each_Pang;@2022-Bacterial_Lee], but these are mainly relied on the plasmid-transferred genes, few phage contributed HGT were reported previously on the topic of long-chain alkane degradation.

The sequences of these AHs were annotated and compared within each group. After screening the full set of viral fragments that coding AlkB, we observed the amino-acid identity of AlkB is between 30%-50% when compared with the reference bacteria genome (Fig 2A), when the protein families of AlkB-phages were evaluated and compared, it’s obvious that AlkB is the only protein that specific to all 6 AlkB-phages (Fig 2A). This suggests that these viral-encoded AlkB were either arising independently or acquired via horizontal gene transfer. A complete AlkB-phage (UViG_3300035703_002088, 719kb with ITR detected) was chosen to show the genome assembling pattern (Fig 2B), this huge phage showing low coding density and abnormal genome assembling pattern, may be due to that some large genome phages use alternative codes and reassign some of the stop codons to be translated as amino acids[@2022-Widespread_Borges;@2020-Clades_Al-Shayeb]. We observed the *AlkB* frequently located in the beginning or ending part of the phage scaffolds(Fig 2B, Fig S1), and besides *AlkB*, almost all the AH genes in this study was located where switch strands happened (Fig S2,Fig S3), showing the mosaicity of phage genome that related to the different origin of the genes[@2020-Phage_Dion;@2019-PHANOTATE_McNair]. 

When the CYP153-phages’ genome were compared, relatively higher amino acid identity between 50-95% were observed (Fig 2C). What’s more, the similarity between different contigs were shown after scaffold alignment, UViG_3300020463_000001 and UViG_3300020438_000013 were highly similar in the compared fragments of genome, same for UViG_3300002092_000034 and UViG_3300032239_000241 (Fig 2C). The former 2 contigs were both isolated from marine ecosystem, while the latter 2 were from fresh water (Fig 2C). showing the possibility that 2 in-depended evolutionary events were existing when CYP153 were acquired by phages. the complete phage genome of UViG_3300020463_000001 (200Kb, DTR detected)were used here to check the details, which shown several tRNA copies in the genome (Fig 2D), notably, 4 other CYP153-coding phage also encode multiple tRNAs (Fig S2), which might contribute to avoid bacterial defence and compensate the differences in codon usage between phages and their bacteria hosts [@2007-Causes_Bailly-Bechet;@2014-Aeromonas_Prabhakaran]. these tRNA-abundant phages indicating their potential active life cycles as the translation is under self-regulation.

One HQ LadA was observed in the viral genome, which is assembled from a bioreactor wastewater sample. Despite the rare incidence, LadA in phage is with high noisy index and when compare the viral genome with a bacteria reference, LadA is the only protein shared between the phage and reference (Fig 2E). LadA homologues was firstly identified in *G. thermodenitrificans* NG80-2 as a single copy on the plasmid pLW1071 [@2007-Genome_Feng], suggest that its an alien gene, later in *G. thermoleovorans* B23 chromosome, a “*LadAB* gene island” was observed that consisted of 3 LadA homologues [@2014-Cloning_Boonmak], and all these 3 homologues are proved to be functional. The amino acid similarity of LadAα in phage is 35%  (Fig 2E). Although the MAG is relative short in the case of LadA-phage(~20kb, Fig 2F), this contig coding another AMG in the genome that annotated by Phrog as acyl-CoA N-acyltransferase, which is important for lipid metabolism [@2015-AcylCoA_Fu].

Two contigs with similar genome were screened from the data base as AlmA-phage(Fig 2G), both were from river water and with DTR detected, they were identified as complete genomes, the size is exactly the same of 364,247 bp (table S1). UViG_3300048625_000166 were chose to show the assembled genome, it’s interesting that similar with the large contig of CYP153-phage, copies of tRNA were observed (Fig 2H), and besides, this phage coding other AMGs as well, and for all the high quality phages in this studies, multiple copies of AMGs could be observed in the genome (Fig S2).

### vHYDEGs have conserved protein structures and active catalytic site

Viruses usually have low protein sequence similarities when compared with homology genes from their host, thus the sequence-based comparison usually failed to detect remote homologous genes. However, protein structures of homologous genes were usually more conserved than their sequences. So, in order to validate that vHYDEGs have the same metabolic potential as their host’s homologues, we used a structural-based comparison approach. Comparing vHYDEG protein 3D structures ensures that viruses-encoded proteins has the necessary domain for enzymatic reactions. We used the state-of-art computational methods predicted protein structures of vHYDEGs, and compared them with those experimentally determined reference structures from RCSB PDB database or Alphafold2 predicted structures from UniProt database to find out most similar structures. Binding sites were highlighted on the 3D structures if they can be obtained from the UniProt feature annotations. Results showed that phage-encoded vHYDEGs has high structural similarity with those reference structures. Binding sites regions of AlkB, AlmA and LadA were overlapped with those in the reference proteins. CYP153, pBmoA and pBmoC has no binding sites annotation in the reference database, but their TM-scores are from 0.83-0.94, suggesting they have the same structures with their reference proteins (Fig 3). These results indicating that phage-encoded HYDEGs maintained the functional domain and might be active when infecting host. While AlkB, AlmA and CYP153 are full-length proteins, LadA only has alpha subunit, and pBmoA and pBmoC are only partial of the pMMO genes. This indicates that viruses usually only encode partial of the full gene. This findings is the same as previous study that show viruses only encode the core domains of a large enzymes [@2020-Large_Chen]. Although structural comparisons suggest vHYDEGs maintained conserved structures of their reference proteins, further experimental evidence is needed to validate their hydrocarbon-degradation potential.

### Diversity and evolution of the vHYDGEs

Phylogenetic analysis of these vHYDEGs showed they are related to diverse bacterial species (Fig 4), but majority of which belong to phylum Proteobacteria. For the AlkB-phages, the amino acid sequences from 7 mVCs formed 3 relatively distant phylogenetic groups, closely related to Bacteroidota, Proteobacteria, and Candidatus Blackallbacteria, respectively (Fig 4A). Many bacterial species reported to be oil degraders belong to the first two phyla [@2018-Petroleum_Xu;@2022-Degradation_Lyu], while the recent-discovered uncultivated phyla Candidatus Blackallbacteria, which were identified from CO2-derived geyser, have not yet reported to be related with long-chain alkane degradation[@2018-Differential_Probst].

For CYP153-phages, they were grouped into 2 separated cluster in the phylogenic tree (Fig 4B), which resemble the genome scaffold alignment(Fig 2C). The two clusters were closely related to Actinobacteria and Proteobacteria, and most of these related bacteria were able to survive in oil contaminated habitat such as *Pseudomonas*, *Rhodococcus* was observed in Arabian gulf sediment [@2020-Biodegradation_Al-Thukair], also *Acinetobacter baumannii* from ACB complex were reported to co-metabolite crude oil with *Talaromyces sp* [@2021-Combined_Zhang]. These evolutionary closeness revealed the possibilities of HGT between these oil degradation bacteria and their phages, although only in rare case. The mVCs with complete protein length of CYP 153, however, is more abundant compared with other 3 AHs, the potential explanation might be firstly the length of CYP153 is shortest among 4, so the possibility of phage to carry and maintain it is higher than the others, and secondly both AlkB and LadA are frequently located in the OCT plasmid or pLW1071 plasmid of bacteria[@2011-Alkanedegrading_Alonso-Gutierrez;@2022-Genomeresolved_Somee], thus the horizontal transfer might more likely happened by plasmid transfer.  

LadA and AlmA are similar as both of then are flavin-dependent monooxygenase of the family of bacterial luciferase, while the structure of LadA is quite unique, 3 functional domains were segregated and 1 act as monooxygenase, other 2 for NADPH oxidation. This structure enables LadA to hydroxylate alkanes without the rubredoxin and rubredoxin reductase, which are required for other AHs enzyme systems[@2008-Crystal_Li]. In the phylogenic tree of LadA(Fig 4C), only 1 of the LadAα  homologues in virus were shown, and was closely related to Rhodospirillales and other Proteobacteria. It’s interesting as in the previous studies that identified LadA in bacteria, they are mainly taking about Geobacillus and Aeribacillus[@2016-Detection_Tourova;@2014-Cloning_Boonmak], which belong to Firmicutes. LadA homologues have also been discovered in fungi[@2022-Identification_Perera]. The metabolic gene clusters in fungi was also hypothesized to be a result of HGT[@2018-Birth_Rokas], earlier work also shown that the KilA-N domain which was firstly identified in bacteriophage P1, was later shown as endogenized viral genes in the genome of bacteria and fungi, all these evidence highlighting the HGT as a mechanism of genetic innovation in eukaryotes as well[@2008-Massive_Gladyshev;@2019-Evolutionary_Medina]. 

Two AlmA-phages phylogenically clustered together and the amino acid is 42% identical with bacterial reference (Fig 2G), the two most popular genera that hosting *almA* gene are *Alcanivorax* and *Marinobacteria*[@2012-Diversity_Wang], and the latter one shown in the phylogenic relationship with the viral-coding AlmA (Fig 4D). 

Both PmoC and PmoA are both subunits of pMMO, which is involved in methane oxidation[@2020-Large_Chen]. PmoA is phylogenically related with methanotrophs such as Methylomagum, Methylocadum and Methylococcus(Fig 4E), indicating the potential co-evolvement of phage and host bacteria. However, the PmoC is closely related with nitrogen cycling bacteria that in Proteobacteria(Fig 4F). This result resemble the phage-coding *amoC* genes from marine samples [@2020-Potential_Gazitua], and we also observed that the protein structure of AmoC is that study is quite similar with our PmoC(Fig 3F), this interesting result showing the potential multiple function of pmoC/amoC genes in phages. These findings suggest that phages can potentially provide new capabilities to a microbial community.

### mVCs coding vHYDEGs are taxonomically diverse

Taxonomic network showed that HYDEG-encoding viruses are mainly mainly distributed in the order *Caudovirales*, and arranged into three main clusters with reference *Myoviridae*, *Siphoviridae* and *Podoviridae*, and three individual viruses were arranged outside main clusters. Of the four viral contigs with ambiguous/unknown predictions, three were clustered with *Myoviridae* and *Siphoviridae*, suggesting that most of of HYDEG-encoding viruses are closely related to *Caudovirales* families (Fig 5). The order *Caudovirales* is the most abundant order of phages in the environment and includes phages that infect a wide range of hosts, including Gram-positive and Gram-negative bacteria, archaea, and even eukaryotes. In addition, some viruses belongs the the class of M*egaviricetes*. *Megaviricetes* are a group of viruses that infect eukaryotic hosts and are characterized by a large genome size and a complex life cycle. The most studied viruses in this group are the *Mimiviridae*, which infect amoebae and are known for their ability to evade host defenses. We detected one *Miniviridae* which encode AlkB. Other viruses that encode AlkB are mainly *Podoviridae,* which infect bacteria and are known for their ability to degrade the bacterial cell wall. The *Podoviridae* are the most abundant phages in the environment are responsible for the majority of phage-host interactions. Most of HYDEG-encoding viruses are clustered together in the network, while some are located in different cluster, indicating that they might have different evolutionary history or have acquired HYDEG using different mechanisms.

### Viruses that encode vHYDEGs have multiple ecological patterns

In order to study the environmental sources of viruses encoding vHYDEGs, we conducted a thorough investigation of the IMGVR database's genome metadata. Our findings revealed that these viruses were diverse and could be found in a variety of environments, including aquatic, terrestrial, bioreactor, plant, and host-associated samples. While most of these viruses were discovered in freshwater and marine samples, there were a few cases found in extreme environments, such as shale gas reservoirs, saline, and thiocyanate environments (Fig 6A). For example, there are two viruses that encode HQ AlkB were isolated from shale gas reservoirs (Fig 4A), and one encode AlkB from anaerobic-aerobic thiocyanate-remediating bioreactor. The only one LadA alpha was from a wastewater bioreactor. All other viruses encoding CYP153, AlmA, pBmoA and pBmoC were from other Aquatic samples. One of the possible explanations for the rare cases identified from extreme environments could be attributed to the fact that there are fewer public database samples available from these extreme environments as compared to freshwater and marine samples. This indicates that the current database only provides a limited view of the actual diversity of viruses that encode vHYDEGs. We anticipate that more viral metagenomics samples from more diverse environment were sequenced in the future would help to expand current knowledge on vHYDEGs.

Among all the viral contigs that encode HQ and MQ vHYGEDs, four of them have complete genome sequence which determined by DTR or ITR, 29 were proviruses, 8 were giant virus metagenome-assembled genomes (GVMAG), and 228 were linear genomes. Both virulent and temperate viruses are included, indicating diverse viruses could encode various type of vHYDEGs that participate in hydrocarbon degradation (Fig 6B).

## Discussion

Viruses are the most prevalent and vast entities in the biology world. They also encoded the most diverse genomic content which have great influence in other organisms[@2020-Phage_Dion]. Beyond encoding viral-structure related genes, more and more recent studies showed that they can also encode various metebolic genes that may have influence on their host metabolism[@2021-Ecology_Kieft;@2011-Phage_Thompson], and thus play key roles in microbial evolution, marine nutrient cycling and human disease. Since viruses act as vector of horizontal gene transfer, these genes can be transfered to other organisms and influence their metabolism and community, thus will have influence on the the entire biology ecosystem. As a results, study function, especially metabolic functions of viral-encoded genes is of great importance, not only for environment, but also for human health and medicine.

In this study, we identified plenty virus-encoded hydrocarbon-degradation genes, which can participate in the initiation of degradation of various hydrocarbons, including methane, mid-chain and long-chain alkanes, and aromatic hydrocarbons. The genomic content, evolutionary relationship, and protein structural analyses validated their putative hydrocarbon-degradation capability. Considering the sources of geography location, these finding suggest that viruses-encoded hydrocarbon degradation genes might have impact on earth carbon and energy recycling, oil contamination degradation, and so on. The identified vHYDEGs were mainly belong to aerobic degradation of alkanes, and anaerobic degradation genes were identified in some low quality vHYDGEs. This might because less mVCs were assembled from anaerobic environment compared with the aerobic ones. Another possible reason is that, there might be more anaerobic degradation genes that were not been discovered, thus not be included in the reference protein database. It is noteworthy most viruses that encode vHYDEGs were not similar to viral genomes in the NCBI viral refseq database. Considering that most viruses in the NCBI viral refseq database were isolated on known host, this finding indicates there are more diverse viruses in the environment that haven’t be identified, and their role and contribution to the entire ecosystem is yet to be understand.

All of the viruses in this study are from IMG/VR database, which contains viruses identified from mainly metagenomic studies. Due to the limitation of current virus detection tools, it might be possible that some non-viral sequences were classified as viruses and be included in public database. To make sure viral genomes used in this study are really viruses, we used a variety of approach to reduce false positive rate. We removed proviruses and contigs with contaminations.

Accurate functional annotation of proteins is challenging, since different database and annotation method could have different accuracy and sensitivity, especially for viral-encoded genes. To provide a robust annotation, we first used a small hydrocarbon-degradation gene database[@2022-CANTHYD_Khot] to annotate all viral-encoded genes with a relaxed threshold to increase detection sensitivity. Due to the small size of the database, it saves a lot time to screening millions of target proteins. Then we annotated the detected genes using multiple database to get a comprehensive function annotation. The annotation results were classified into multiple confidence level based on annotation score. This strategy significantly increased the detection sensitivity, as well as reduced the false positive rate. One limitation of CANT-HYD database is that it only contain enzymes that participate the beginning of hydrocarbon-degradation pathway. Existing studies already observed disproportionately abundant of phage-coding AMGs in the carbon metabolism, including energy production through PPP and TCA cycle [@2016-Viral_Hurwitz;@2013-Metabolic_Hurwitz;@2016-Ecogenomics_Roux], and 9 glycoside hydrolase families that could degrade pectin, hemicellulose and starch [@2018-Hostlinked_Emerson]. Presence of vHYDEGs in phages now add one previously missing piece into the whole map of  carbon degradation, indicate that phages can potentially provide new degradation possibilities to a microbial community. An explanation for this is that phages with vHYDEG participate the initiation of hydrocarbon activation may provide more evolution advantage for phages than those encode enzymes involved in the middle or end of the degradation pathway, as the AMGs that contributes to critical, rate-limiting steps were thought to helps the phages to boost the host metabolism during infection, which in turn benefit to phage propagation [@2007-Exploring_BREITBART]. Similar observation could be found in phage-encoded *pmoC* gene that is involved in the beginning of the methane degradation pathway [@2020-Large_Chen], and genes participate various sulfur metabolic process [@2021-Ecology_Kieft].

Because of the rapid evolution of virus genes, phages show impressive genomic diversity and complex evolutionary relationships that do not follow traditional hierarchical phylogeny, due to pervasive mosaicism. virus-coding AMGs have less similarity to their homology genes in their host, thus we used a two-step strategy that is able to detect remote-homologous proteins. A shortage of the remote-homologous detection method is that it might introduce many false positive results. To solve this issue, we used a various approach to ensure the identified proteins are really homologous to the query protein. We validated several cases using phylogenetic analysis, genomic content analysis, and protein structure analysis. The diversity of homologous results suggests that this approach is reliable.

We created a database contains all identified putative vHYDEGs with their comprehensive annotations, and link to external resources such as IMG/VR, NCBI and PHROGs database. Researchers can search their interested vHYDEGs in this database, and get the overview of genomic content, as well as sources of the viruses and their host information. Another limitation is that although we tried to minimize false positive hit, there might still be many false positives that were  annotated as HYDEGs. This prompted us to identify functional proteins by predicting their 3D structures using the state-of-art tools like Alphafold2, RoseTTAFold or ESMFold. Structure differences identified in vHYDEGs will give us new insight into the catalytic mechanisms of the enzyme, and provide information for de novo design more efficient enzymes using state-of-the-art technologies [@2023-Novo_Yeh]. Researchers will be able to design more efficient enzymes to degrade hydrocarbons, which can be used in crude oil contamination removal, oil well, and etc.

## Conclusions

In this study, a total of ~5.6 million mVCs were retrieved from IMG/VR databases and less than 0.07% (595) of them encoded at least one HYDEG. We chose 57 high quality mVCs, distinguishing them from proviruses, provide the genome structure of and complete annotation of the full set of genes. These 57 mVCs were proved to coding key-step-enzymes of hydrocarbon degradation, as well as other important genes that involved in protein translation, sulfur and lipid metabolism. To further validate the function of the observed vHYDEGs, the protein structure and catalytic activity site were predicted, showing high-level of structural identity when compared with experimental validated homologous.

The habitat might contribute to the relative independent evolutionary events of the vHYDGs, viruses that coding these gene could be found from several different environments, with diverse taxonomy, vHYDEGs evolutionarily related with collective of bacteria from varied phylum, and clustered into distinct clades.  considering the unexplored environmental samples that out of our current reach, it’s reasonable to hypothesis that there are novel enzymes participate hydrocarbon degradation pathway, but hasn’t be recorded in public protein database. Database contains current available viruses-encoded hydrocarbon degradation genes may present a reservoir of novel nucleic acid manipulation tools in genome editing and might benefit future research.

### Data and code availability

All sequences and custom analysis scripts used in this study are available at [https://github.com/deng-lab/vhydeg](https://github.com/deng-lab/vhydeg). The vHyDeg database is available at [https://deng-lab.github.io/vhydeg](https://deng-lab.github.io/vhydeg).


### References

::: {#refs}
:::


### Figures

![Fig 1](`r here(path_data, "1.png")`)

![Fig 2](`r here(path_data, "2.png")`)

![Fig 3](`r here(path_data, "3.png")`)

![Fig 4](`r here(path_data, "4.png")`)

![Fig 5](`r here(path_data, "5.png")`)

![Fig 6](`r here(path_data, "6.png")`)
